### PR TITLE
Add a flag for allowing single quotes in JSON strings. 

### PR DESCRIPTION
**Pull request by nvdbaranec**
Add a flag that allows `get_json_object()` to accept JSON with strings using single quotes.   Also adds an explicit `get_json_object_options` struct for allowing the user to customize what behaviors they want.   

Note:  stripping quotes from individually returned string values has been left _on_ as default.

### CODE REVIEW

File: cpp/include/cudf/strings/json.hpp [hunk 1]
```diff
@@ -26,6 +28,78 @@ namespace strings {
  * @file
  */
 
+/**
+ * @brief Settings for `get_json_object()`.
+ */
+class get_json_object_options {
+  // allow single quotes to represent strings in JSON
+  bool allow_single_quotes = false;
```
Comment (davidwendt) [inline cpp/include/cudf/strings/json.hpp:36]: Why should this be an option? When would we not want to support both single-quotes and double-quotes?

**Comment by nvdbaranec** [comment]
Pinging @jlowe for any input on the flipping of the logic (previously, default behavior was Spark-specific, now it is the opposite and the plugin will need to set `set_allow_single_quotes(true)` and `set_strip_quotes_from_single_strings(true)` .

File: cpp/include/cudf/strings/json.hpp (hunk 1 above)
Comment (nvdbaranec) [reply cpp/include/cudf/strings/json.hpp:36]: Single quotes is actually not valid JSON.  This is a weird standard violation.

Comment (ttnghia) [reply cpp/include/cudf/strings/json.hpp:36]: Why don't throw an exception when there is a single quote? Then the users can know that they were doing wrong and will fix their data, instead of continuing to use wrong data.

Comment (davidwendt) [reply cpp/include/cudf/strings/json.hpp:36]: The cuIO JSON parser does not have this option. I think the caller should be required to pass correct JSON.

Comment (nvdbaranec) [reply cpp/include/cudf/strings/json.hpp:36]: That's not really an option.  This is something Spark and Hive explicitly support.

Comment (davidwendt) [reply cpp/include/cudf/strings/json.hpp:36]: Can you not do a `replace("\'", "\"")` first?

Comment (ttnghia) [reply cpp/include/cudf/strings/json.hpp:36]: Yeah, can we do one such replacement then process the result as there are only double quotes?

Comment (jlowe) [reply cpp/include/cudf/strings/json.hpp:36]: Won't that break properly formed JSON string values that contain a single-quote character, e.g.: `"this is a 'test' string"`. Even if it somehow doesn't break the parsing, it will munge the string's payload value in an unrecoverable way, correct?

Comment (ttnghia) [reply cpp/include/cudf/strings/json.hpp:36]: I think the replacement is not just simple as `replace("\'", "\"")`, but can be done by following some pattern, such as checking the single quote and replace only if it is at the beginning of the string etc.

Comment (davidwendt) [reply cpp/include/cudf/strings/json.hpp:36]: The single-quote version `'this is a 'test' string'` should not exist if the flag is on as well. 
So it should be safe to just replace all single-quotes with double-quotes when you know the flag is to be used.

Comment (nvdbaranec) [reply cpp/include/cudf/strings/json.hpp:36]: > The single-quote version `'this is a 'test' string'` should not exist if the flag is on as well.
> So it should be safe to just replace all single-quotes with double-quotes when you know the flag is to be used.

That's not correct. Here's some valid JSON:

`{ "a" : "This is a 'test' "}`

If we instead had 

`{ 'a' : 'This is a 'test' '}`

Doing a replace on ' would turn it into

`{ "a" : "This is a "test" "}`

We definitely do not want this.

Regardless, even if we did a replace thing, it would be purely worse from a performance perspective.  We'd have to allocate a potentially massive intermediate buffer (the whole string column) just to do it.   That seems pretty wasteful just to avoid a bool option.

Comment (davidwendt) [reply cpp/include/cudf/strings/json.hpp:36]: Even with the single-quote flag, this will fail to parse: `{ 'a' : 'This is a 'test' '}` right?
So, if this cannot exist, then a replace will work.
If you control the column data  inside the Spark plugin you could do the replace in-place since its just one-byte exchanged for another.

So, maybe this should not be a libcudf function. I think we should be concerned with inconsistencies with cuIO.
@harrism @kkraus14 @jrhemstad

Comment (nvdbaranec) [reply cpp/include/cudf/strings/json.hpp:36]: > So, maybe this should not be a libcudf function. I think we should be concerned with inconsistencies with cuIO.

To reiterate, this behavior is _off_ by default.  The default behavior would match cuIO.

Comment (kkraus14) [reply cpp/include/cudf/strings/json.hpp:36]: For what it's worth, my intuition is that it will only be a matter of time until someone asks us to support single quotes in the cuIO JSON reader as well. Unfortunately, this is a situation that exists in the wild world of ETL and isn't something that can be easily worked around via string replacements.

File: cpp/include/cudf/strings/json.hpp [hunk 2]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2019-2021, NVIDIA CORPORATION.
+ * Copyright (c) 2021, NVIDIA CORPORATION.
```
Comment (hyperbolic2346) [inline cpp/include/cudf/strings/json.hpp:2]: Was this change intended?

File: cpp/src/strings/json/json_path.cu [hunk 3]
```diff
@@ -386,11 +400,17 @@ class json_state : private parser {
     return parse_result::SUCCESS;
   }
 
+  CUDA_HOST_DEVICE_CALLABLE bool is_quote(char c)
+  {
+    return options.get_allow_single_quotes() ? (c == '\"' || c == '\'') : c == '\"';
```
Comment (davidwendt) [inline cpp/src/strings/json/json_path.cu:405]:
```suggestion
    return (c == '\"') || (options.get_allow_single_quotes() && (c == '\''));
```
Slightly less instructions.

File: cpp/src/strings/json/json_path.cu [hunk 4]
```diff
@@ -327,15 +290,62 @@ class json_state : private parser {
...
+        return parse_result::SUCCESS;
+      }
+    }
+    return parse_result::EMPTY;
+  }
+
  private:
+  // numbers, true, false, null.
+  // this function is not particularly strong. badly formed values will get
+  // consumed without throwing any errors
```
Comment (davidwendt) [inline cpp/src/strings/json/json_path.cu:318]: Make this into a doxygen style comment that also includes a description for the `val` parameter.

File: cpp/src/strings/json/json_path.cu [hunk 5]
```diff
@@ -327,15 +290,62 @@ class json_state : private parser {
     return parse_result::ERROR;
   }
 
+  // a name means:
+  // - a string followed by a :
+  // - no string
+  CUDA_HOST_DEVICE_CALLABLE parse_result parse_name(string_view& name, bool can_be_empty)
```
Comment (davidwendt) [inline cpp/src/strings/json/json_path.cu:296]: Doxygen style comment here too.

File: cpp/include/cudf/strings/json.hpp [hunk 6]
```diff
@@ -26,6 +28,76 @@ namespace strings {
...
+  /**
+   * @brief Default constructor.
+   */
+  explicit get_json_object_options() = default;
+
+  /**
+   * @brief Returns true/false depending on whether single-quotes for representing strings
+   * are allowed.
+   */
+  CUDA_HOST_DEVICE_CALLABLE bool get_allow_single_quotes() const { return allow_single_quotes; }
```
Comment (ttnghia) [inline cpp/include/cudf/strings/json.hpp:51]: To be honest, options struct with getters/setters make no sense to me. Can we use just struct like this?
```
struct options {
  // ...
  bool allow_single_quotes = false;

  //...
  bool strip_quotes_from_single_strings = false;
};
```

**Review by ttnghia — CHANGES_REQUESTED** [review]

File: cpp/src/strings/json/json_path.cu [hunk 7]
```diff
@@ -217,19 +177,23 @@ class json_state : private parser {
...
-  __device__ json_state(const char* _input, int64_t _input_len)
+  __device__ json_state(const char* _input,
+                        int64_t _input_len,
+                        thrust::optional<get_json_object_options> _options)
     : parser(_input, _input_len),
       cur_el_start(nullptr),
       cur_el_type(json_element_type::NONE),
       parent_el_type(json_element_type::NONE)
   {
+    options = _options.value_or(get_json_object_options{});
```
Comment (ttnghia) [inline cpp/src/strings/json/json_path.cu:188]: You pass an optional object which a default value, then you also get the object from the optional with a default fallback. Can we totally get rid of optional and use a default object at the beginning?

File: cpp/include/cudf/strings/json.hpp (hunk 6 above)
Comment (nvdbaranec) [reply cpp/include/cudf/strings/json.hpp:51]: I tend to like what you're proposing, but the getter/setter model is consistent with the rest of cudf (cuio uses this a lot)

File: cpp/include/cudf/strings/json.hpp [hunk 8]
```diff
@@ -26,6 +28,76 @@ namespace strings {
  * @file
  */
 
+/**
+ * @brief Settings for `get_json_object()`.
+ */
+class get_json_object_options {
```
[the remark this answers is not in the document]
Comment (nvdbaranec) [reply cpp/include/cudf/strings/json.hpp:34]: I would normally agree, but cuIO has a JSON parser and using a less descriptive name might be unclear/confusing.

**Comment by nvdbaranec** [comment]
As per Paul's comment, https://github.com/rapidsai/cudf/pull/8144#issuecomment-834518807 I have switched the default behavior for stripping quotes on single-string return values back to the way they were.   Note:  allowing single-quotes is still _off_ by default, so the Spark plugin will need to toggle this on.

File: cpp/tests/strings/json_tests.cpp [hunk 9]
```diff
@@ -759,3 +759,109 @@ TEST_F(JsonTests, MixedOutput)
...
+    "}",
+
+    "{"
+      "\"a\": {"
+                "\'z\': {\'i\': 10, \'j\': 100},"
+                "\'b\': [\'c\',null,true,-1]"
+              "}"
+    "}"
+  };
+  // clang-format on
```
Comment (davidwendt) [inline cpp/tests/strings/json_tests.cpp:839]:
```suggestion
    // clang-format on
  };
```

File: cpp/tests/strings/json_tests.cpp [hunk 10]
```diff
@@ -759,3 +759,109 @@ TEST_F(JsonTests, MixedOutput)
...
+{
+  // Tests allowing single quotes for strings.
+  // Note:  this flag allows a mix of single and double quotes. it doesn't explicitly require
+  // single-quotes only.
+
+  // various queries on:
+  // clang-format off
+  std::vector<std::string> input_strings {
+    "{\'a\': {\'b\' : \'c\'}}",
+
```
Comment (davidwendt) [inline cpp/tests/strings/json_tests.cpp:814]: Perhaps include a test string like ` "abc'def" `.
Otherwise, a simple global replace of `'` with `"` before calling `get_json_object` would work for this.

**Review by davidwendt — CHANGES_REQUESTED** [review]

File: cpp/src/strings/json/json_path.cu [hunk 11]
```diff
@@ -327,15 +288,86 @@ class json_state : private parser {
...
+  /**
+   * @brief Parse a name field for a JSON element.
+   *
+   * When parsing JSON objects, it is not always a requirement that the name
+   * actually exists.  For example, the outer object bounded by {} here has
+   * no name, while the inner element "a" does.
+   *
+   * {
+   *    "a" : "b"
+   * }
```
Comment (davidwendt) [inline cpp/src/strings/json/json_path.cu:300]:
```suggestion
   * ```
   * {
   *    "a" : "b"
   * }
   * ```
```

File: cpp/src/strings/json/json_path.cu [hunk 12]
```diff
@@ -327,15 +288,86 @@ class json_state : private parser {
...
+   *
+   * The user can specify whether or not the name string must be present via
+   * the `can_be_empty` flag.
+   *
+   * When a name is present, it must be followed by a :
```
Comment (davidwendt) [inline cpp/src/strings/json/json_path.cu:305]:
```suggestion
   * When a name is present, it must be followed by a colon `:`
```

File: cpp/src/strings/json/json_path.cu [hunk 13]
```diff
@@ -327,15 +288,86 @@ class json_state : private parser {
...
+   *
+   * @param name (Output) The resulting name.
```
Comment (davidwendt) [inline cpp/src/strings/json/json_path.cu:307]:
```suggestion
   * @param[out] name The resulting name.
```

File: cpp/src/strings/json/json_path.cu [hunk 14]
```diff
@@ -97,80 +98,39 @@ class parser {
     return false;
   }
 
+  // if quote = 0, allow either ' or "
```
Comment (davidwendt) [inline cpp/src/strings/json/json_path.cu:101]: Can doxygen-ify this one too?

**Review by davidwendt — APPROVED** [review]

File: cpp/include/cudf/strings/json.hpp [hunk 15]
```diff
@@ -17,6 +17,8 @@
 
 #include <cudf/strings/strings_column_view.hpp>
 
+#include <thrust/optional.h>
```
Comment (hyperbolic2346) [inline cpp/include/cudf/strings/json.hpp:20]: Should we convert this to std::optional now that we support c++17?

**Review by hyperbolic2346 — APPROVED** [review]
Looks good to me

File: cpp/include/cudf/strings/detail/json.hpp [hunk 16]
```diff
@@ -32,6 +32,7 @@ namespace detail {
 std::unique_ptr<cudf::column> get_json_object(
   cudf::strings_column_view const& col,
   cudf::string_scalar const& json_path,
+  get_json_object_options options     = get_json_object_options{},
   rmm::cuda_stream_view stream        = rmm::cuda_stream_default,
```
Comment (ttnghia) [inline cpp/include/cudf/strings/detail/json.hpp:36]: Typically I don't set default parameters for the `cudf::detail::` API, because that can lead to confusion. You can call the `cudf::` and `cudf::detail::` versions by the same list of arguments. For example: 
```
cudf::strings::get_json_object(col, json_path);
cudf::strings::detail::get_json_object(col, json_path);
```

So, if possible, please remove default parameters. By doing so, if you want to call `get_json_object(col, json_path)` then you can call the `cudf::strings::` version, and if you want to call `get_json_object(col, json_path, options,...)` then you must call the `cudf::strings::detail::` version.

File: cpp/src/strings/json/json_path.cu [hunk 17]
```diff
@@ -97,80 +98,48 @@ class parser {
...
+   * @brief Parse a quote-enclosed JSON string.
+   *
+   * @param[out] str The resulting string.
+   * @param can_be_empty Parameter indicating whether it is valid for the string
+   * to not be present.
+   * @param quote Character expected as the surrounding quotes.  A value of 0
+   * indicates allowing either single or double quotes (but not a mixture of both).
+   * @returns A result code indicating success, failure or other result.
+   */
   CUDA_HOST_DEVICE_CALLABLE parse_result parse_string(string_view& str,
```
Comment (ttnghia) [inline cpp/src/strings/json/json_path.cu:111]: I'm not sure if it is too late to comment here: This C-style function call (returns error code while the output is a reference parameter) should be avoided. Now we can use optional or something like that.

File: cpp/include/cudf/strings/json.hpp (hunk 15 above)
Comment (ttnghia) [reply cpp/include/cudf/strings/json.hpp:20]: Agree 😄

**Review by ttnghia — APPROVED** [review]

File: cpp/src/strings/json/json_path.cu (hunk 17 above)
Comment (harrism) [reply cpp/src/strings/json/json_path.cu:111]: You can use exceptions instead of error codes.

File: cpp/include/cudf/strings/detail/json.hpp (hunk 16 above)
Comment (nvdbaranec) [reply cpp/include/cudf/strings/detail/json.hpp:36]: Done

**Comment by nvdbaranec** [comment]
Changing this to non-breaking as it won't alter existing behavior.

**Comment by nvdbaranec** [comment]
@gpucibot merge
